### PR TITLE
[SofaQtQuick] Pass extra command-line arguments for python scenes in a more high-level function call

### PR DESCRIPTION
**Pull request by Ant0nin**
Pass extra command-line arguments for python scenes loading in a more high-level function call. Knowing that SofaQtQuick uses a different argument parser, the underlying objective of this commit is to be able to pass command-line extra args using QCommandLineParser. This feature is ready on the side of SofaQtQuick and it works (i.e. attached patch). Note that the added input parameter 'sceneArgs' in the 'load' functions is an optional parameter.

[patch_pyargs_support.txt](https://github.com/sofa-framework/sofa/files/3073394/patch_pyargs_support.txt)

______________________________________________________

This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by epernod** [comment]
[ci-build][with-all-tests]

**Comment by guparan** [comment]
Thanks for this PR @Ant0nin. We need a bit more info though. Why do you need to forward sceneArgs through all loading functions (deep) that actually do nothing with it? Why can't you use `sofa::helper::ArgumentParser::extra_args()`?

**Comment by Ant0nin** [comment]
@guparan In fact, calling `sofa::helper::ArgumentParser::extra_args()` from `SceneLoaderPY::doLoad` (line 89 of SceneLoaderPY.cpp) causes an inconvenient limitation: The python scene loader cannot be used correctly when using a different parser than `sofa::helper::ArgumentParser`. 

And in SofaQtQuick, we precisely would like to use the class `SceneLoaderPY` but using `QCommandLineParser` (`sofa::helper::ArgumentParser` is not used at all in SofaQtQuick).

In the suggested commit, I simply move the call of `sofa::helper::ArgumentParser::extra_args()` from `SceneLoaderPY::doLoad` into more high-level methods. But this change induces to add a second input parameter for `SceneLoaderPY::doLoad`, in order to forward sceneArgs. Knowing that `doLoad` is originally a pure virtual method defined in class SceneLoader (see file SceneLoaderFactory.h), the multiple method overrides of this one  (so in each scene loader) have to take into account the sceneArgs parameter even if it does nothing with it, that's why I made the parameter optional (NULL by default).

File: SofaKernel/framework/sofa/simulation/SceneLoaderFactory.h [hunk 1]
```diff
@@ -66,14 +66,14 @@ class SOFA_SIMULATION_CORE_API SceneLoader
     virtual bool canWriteFileExtension(const char * /*extension*/) { return false; }
 
     /// load the file
-    sofa::simulation::Node::SPtr load(const char *filename)
+    sofa::simulation::Node::SPtr load(const char *filename, const std::vector<std::string>* sceneArgs = NULL)
```
Comment (damienmarchal) [inline SofaKernel/framework/sofa/simulation/SceneLoaderFactory.h]: passing a const reference is often a better choice. No risk of null pointer so no corner case to handle in the code. 
The default parameter can be an empty vector of string.

**Review by damienmarchal — COMMENTED** [review]
Hello,

Thanks for the PR. 
- I think the adding of an extra parameter is fine. 
- Use const std::vector<string>& instead of a pointer to it.
- As this is touching the load function I suggest to take profit of the PR to change the function signature from
```cpp
void doLoad(char*,const vector<std::string>...) 
```
to
```cpp
void doLoad(const std::string&, const vector<std::string>...) 
```
Because...well mixing C string & C++ string smells like c++ code from the late 90's.

File: SofaKernel/framework/sofa/simulation/SceneLoaderFactory.h (hunk 1 above)
Comment (Ant0nin) [reply SofaKernel/framework/sofa/simulation/SceneLoaderFactory.h]: I totally agree with you @damienmarchal . I made the asked modifications. Please tell me if it is OK.

**Comment by guparan** [comment]
So we re-discussed this PR yesterday at SOFA-dev meeting and I still struggle to understand why would it be hard to use `sofa::helper::ArgumentParser` in SofaQtQuick.
Could you please re-explain? Maybe @damienmarchal too? Thanks.

**Comment by damienmarchal** [comment]
So,
1) there is no reason to have helper::ArgumentParser in SceneLoaders because a loader (py, scn, etc..) shouldn't be connected in any way to the command line (which is an user interface aspect), especially one relying on a static global object as ArgumentParser. 
2) it make sense to load a scene with specific parameters...so having an array with scene specific parameters is in the load function seems sounded.

DM.

**Comment by damienmarchal** [comment]
@guparan what remains to do on this code ?  I would be nice to have it merge.

**Comment by damienmarchal** [comment]
Any news on this @guparan ?

**Comment by Ant0nin** [comment]
I just rebased the PR branch (command_args_qtquick_compat) onto the current master branch, and solve the conflicts.

**Comment by guparan** [comment]
Sorry for the long silence. I see no objection to move on with this PR :+1:

**Comment by guparan** [comment]
@Ant0nin all python related tests are failing, could you take a look?

**Comment by Ant0nin** [comment]
@guparan I updated the PR branch by merging the master branch into it, and it seems to solve the issue.

**Comment by guparan** [comment]
@Ant0nin The conflicts are indeed solved but the test errors remain. See https://www.sofa-framework.org/dash?branch=origin/PR-992
Could you try for instance to run SofaPython_test on your local build?

**Comment by Ant0nin** [comment]
@guparan you're right! We didn't see the problem because we ran the tests on the Release build and not on the Debug build. A segmentation fault occurs with the Debug build, dues to an assertion. @tgaugry currently investigates on the issue.

**Comment by Ant0nin** [comment]
@guparan This PR has been closed but the branch has not been merged onto master... why?

**Comment by guparan** [comment]
Woops, my bad, it was not intended

**Comment by Ant0nin** [comment]
Thanks a lot for your efforts guys! I know this PR was a pain in the ass to merge and I apologize for this. I am happy it's finally done! Great job!

**Pull request merged**